src/coins.h

```diff
@@ -215,10 +216,11 @@ class CCoinsViewCache : public CCoinsViewBacked
      * declared as "const".
      */
     mutable uint256 hashBlock;
-    mutable CCoinsMap cacheCoins;
+    mutable bulk_pool::Pool cacheCoinsPool{};
```

Comment:
Given that the pool only grows and uses increasing memory over time, and that its lifetime is effectively the lifetime of the entire node, would it make sense to add some kind of garbage collection method for `CCoinsViewCache` that would free unneeded memory (maybe after finishing syncing, or after flushing, or on an interval, or from an RPC a user could call)?

Reply:
I have the following code for something i'm working on. It should be amortized cheap to call after every erase; you could also make it have different parameters during sync or after (e.g, always compact after sync, only compact if it's above 50MB free during sync).
```c++
 static void resize_if_savings(CTxMemPoolEntry::relatives& relatives) {
       assert(relatives.max_load_factor() == 1)
       // If we're at 0, clear out the map
       if (relatives.size() == 0) {
           CTxMemPool::relatives tmp;
           std::swap(tmp, relatives);
           return;
       }
       // don't bother saving for small enough sets
       // 19 buckets isn't very large, and fits in with the usual
       // prime rehash policies
       if (relatives.bucket_count() <= 19) return;
       // don't bother rehashing if we're more than half full
       const size_t full_size = relatives.bucket_count();
       if (relatives.size() > full_size/2) return;
   
       CTxMemPool::relatives tmp{std::make_move_iterator(relatives.begin()), std::make_move_iterator(relatives.end()), relatives.size()};
       std::swap(tmp, relatives);
   }
```

Reply:
@JeremyRubin that looks interesting. I think the `if (relatives.size() > full_size/2) return;` is a bit too aggressive though. For a hashmap uses doubling as resize strategy: if it has e.g. 256 elements and you add one it will double its size, and when you remove one element again it would already trigger the rehashing to decrease its size.

Reply:
On my system the rehash policy is the  prime less than double policy, which ensures that we have a growth factor < 2 for all resizes. So in the case of 256, we would first "double" to 503. Then we would be at 257/503 "fullness", where half would be 251. So we can remove up to 6 elements before triggering a re-hash. So the pathological cases are a bit more precise, see below:


If you're at exactly 251, then you add one, you double to 503, and then you can remove two before triggering a rehash (strict inequality). Let's say you choose to remove a third. Then you'll rehash to either 467 or 257 (implementations depending -- I believe it's usually 257 because we go to the smallest prime larger than N). That means you will need to remove down to 233 (~20 more elements) before triggering a rehash OR more likely ~125 elements. Let's say instead, if you add back another N, you won't trigger a rehash until you hit 467, so you can add back another ~215 elements before having to go again OR more likely, 7 elements.

So there is a case to be made that it's less than ideal, but it's not quite as bad given the primes.

There's also a case to be made (at least for the use cases I'm looking at) where you're usually in a streak of erases, followed by a streak of adds. So while it is possible to trigger the bad behavior, it's atypical.

Therefore I don't actually think it's that bad to use half as the shrink test, given that it quickly gets us shrunk to a more than half full table.

But if we do a resize at, say, at n <= (buckets - (buckets/100)*54), and resize to (n + (n/100)*36), then we can always be able to add at least 36% after a resize and remove 1/(1 + 0.36*1.08) - 0.36 ~~ 36% more before resizing again.  (The ~1.08 more than 36% I think, is half the average overshoot prime to prime. This is not a great estimate but can be improved).